File: plilja-python3/day21/day21.py
import sys

def step1(ops):
    return scramble('abcdefgh', ops)


def step2(ops):
    return unscramble('fbgdceah', ops)


def unscramble(start, ops):
    for i in range(0, len(ops)):
        if 'left' in ops[i]:
            ops[i] = ops[i].replace('left', 'right')
        else:
            ops[i] = ops[i].replace('right', 'left')
        if 'move' in ops[i]:
            parts = ops[i].split()
            parts[2], parts[5] = parts[5], parts[2]
            ops[i] = ' '.join(parts)
        if 'rotate based on' in ops[i]:
            parts = ops[i].split()
            ops[i] = 'rotate special %s' % parts[6]
    return scramble(start, reversed(ops))


def scramble(start, ops):
    def rotate_left(passwd, n):
        n = n % len(passwd)
        return passwd[n:] + passwd[:n]

    def rotate_right(passwd, n):
        n = n % len(passwd)
        return passwd[-n:] + passwd[:-n]

    def rotate_based_on(passwd, c):
        x = passwd.index(c)
        if x >= 4:
            x += 1
        x += 1
        return rotate_right(passwd, x)

    passwd = [x for x in start]
    for op in ops:
        parts = op.split()
        if parts[0] == 'swap' and parts[1] == 'position':
            x = int(parts[2])
            y = int(parts[5])
            passwd[x], passwd[y] = passwd[y], passwd[x]
        elif parts[0] == 'reverse':
            x = min(int(parts[2]), int(parts[4]))
            y = max(int(parts[2]), int(parts[4]))
            for i in range(0, (y - x + 1) // 2):
                passwd[x + i], passwd[y - i] = passwd[y - i], passwd[x + i]
        elif parts[0] == 'swap' and parts[1] == 'letter':
            x = parts[2][0]
            y = parts[5][0]
            for i in range(0, len(passwd)):
                if passwd[i] == x:
                    passwd[i] = y
                elif passwd[i] == y:
                    passwd[i] = x
        elif parts[0] == 'rotate' and parts[1] == 'special':
            c = parts[2][0]
            new_passwd = passwd
            while rotate_based_on(new_passwd, c) != passwd:
                new_passwd = rotate_left(new_passwd, 1)
            passwd = new_passwd
        elif parts[0] == 'rotate':
            if parts[1] == 'based':
                passwd = rotate_based_on(passwd, parts[6][0])
            else:
                x = int(parts[2])
                if parts[1] == 'left':
                    passwd = rotate_left(passwd, x) 
                else:
                    passwd = rotate_right(passwd, x) 
        else:
            assert parts[0] == 'move'
            x = int(parts[2])
            y = int(parts[5])
            tmp = passwd[x]
            passwd = passwd[:x] + passwd[x + 1:]
            passwd = passwd[:y] + [tmp] + passwd[y:]

    return ''.join(passwd)


ops = sys.stdin.readlines()
print(step1(ops))
print(step2(ops))

File: emlun-scala/day21.scala
object Main extends App {

  val swapPositionPattern = raw"swap position (\d+) with position (\d+)".r
  val swapLetterPattern = raw"swap letter (.) with letter (.)".r
  val rotateLeftPattern = raw"rotate left (\d+) steps?".r
  val rotateRightPattern = raw"rotate right (\d+) steps?".r
  val rotateDynamicPattern = raw"rotate based on position of letter (.)".r
  val reverseSlicePattern = raw"reverse positions (\d+) through (\d+)".r
  val moveIndexPattern = raw"move position (\d+) to position (\d+)".r

  def moveIndex(state: String, x: Int, y: Int): String = {
    val without = state.take(x.toInt) ++: state.drop(x.toInt + 1)
    without.take(y.toInt) ++: state(x.toInt) +: without.drop(y.toInt)
  }

  def reverseSlice(state: String, x: Int, y: Int): String =
    state.patch(x, state.slice(x, y + 1).reverse, (y + 1 - x))

  def rotateDynamic(state: String, letter: String): String = {
    val index = state indexOf letter
    val steps = (index + (if (index >= 4) 2 else 1)) % state.length
    state.takeRight(steps) ++ state.dropRight(steps)
  }

  def rotateLeft(state: String, steps: Int): String =
    state.drop(steps.toInt) ++: state.take(steps.toInt)

  def rotateRight(state: String, steps: Int): String =
    state.takeRight(steps.toInt) ++ state.dropRight(steps.toInt)

  def swapLetter(state: String, x: Char, y: Char): String =
    state map {
      case `x` => y
      case `y` => x
      case c => c
    }

  def swapPosition(state: String, x: Int, y: Int): String = {
    val xIndex = x.toInt
    val yIndex = y.toInt
    val (xValue, yValue) = (state(xIndex), state(yIndex))
    state.updated(xIndex, yValue).updated(yIndex, xValue)
  }

  def solve(instructions: List[String], initialState: String): String =
    instructions.foldLeft(initialState) { (state, instruction) =>
      instruction match {
        case swapPositionPattern(x, y)    => swapPosition(state, x.toInt, y.toInt)
        case swapLetterPattern(x, y)      => swapLetter(state, x.head, y.head)
        case rotateLeftPattern(steps)     => rotateLeft(state, steps.toInt)
        case rotateRightPattern(steps)    => rotateRight(state, steps.toInt)
        case rotateDynamicPattern(letter) => rotateDynamic(state, letter)
        case reverseSlicePattern(x, y)    => reverseSlice(state, x.toInt, y.toInt)
        case moveIndexPattern(x, y)       => moveIndex(state, x.toInt, y.toInt)
        case _ => ???
      }
    }

  def solveBackward(instructions: List[String], finalState: String): String =
    instructions.reverse.foldLeft(finalState) { (state, instruction) =>
      instruction match {
        case swapPositionPattern(x, y)    => swapPosition(state, x.toInt, y.toInt)
        case swapLetterPattern(x, y)      => swapLetter(state, x.head, y.head)
        case rotateLeftPattern(steps)     => rotateRight(state, steps.toInt)
        case rotateRightPattern(steps)    => rotateLeft(state, steps.toInt)
        case rotateDynamicPattern(letter) => (
          (0 until state.length)
            map { steps => rotateLeft(state, steps) }
            find { preState => rotateDynamic(preState, letter) == state }
        ).get
        case reverseSlicePattern(x, y)    => reverseSlice(state, x.toInt, y.toInt)
        case moveIndexPattern(y, x)       => moveIndex(state, x.toInt, y.toInt)
        case _ => ???
      }
    }

  val initialStateA = "abcdefgh"
  val initialStateB = "fbgdceah"

  val instructions = (for {
    line <- io.Source.stdin.getLines
    instruction = line.trim
  } yield instruction).toList

  println("A: " + solve(instructions, initialStateA))
  println("B: " + solveBackward(instructions, initialStateB))

}
